pallets/loans/src/types/mod.rs

```diff
@@ -134,6 +166,101 @@ impl RepaymentSchedule {
 	pub fn is_valid(&self, now: Moment) -> bool {
 		self.maturity.is_valid(now)
 	}
+
+	// TODO: this should be exposed through a runtime API by (pool_id, loan_id)
+	pub fn generate_expected_cashflows<Balance, Rate>(
+		&self,
+		origination_date: Moment,
+		principal: Balance,
+		interest_rate: Rate,
+	) -> Result<Vec<(NaiveDate, Balance)>, DispatchError>
+	where
+		Balance: tokens::Balance + FixedPointOperand + From<u128>,
+		Rate: FixedPointNumber,
+	{
+		match self.maturity {
+			Maturity::Fixed(maturity_date) => {
+				let start = NaiveDateTime::from_timestamp_opt(origination_date as i64, 0)
+					.ok_or(DispatchError::Other("Invalid origination date"))?
+					.date();
+				let end = NaiveDateTime::from_timestamp_opt(maturity_date as i64, 0)
+					.ok_or(DispatchError::Other("Invalid maturity date"))?
+					.date();
+
+				match &self.interest_payments {
+					InterestPayments::None => Ok(vec![]),
+					InterestPayments::Monthly(reference_date) => Self::get_monthly_cashflows(
+						start,
+						end,
+						principal,
+						interest_rate,
+						reference_date.clone(),
+					),
+					InterestPayments::SemiAnnually(reference_date) => {
+						Ok(Self::get_monthly_cashflows(
+							start,
+							end,
+							principal,
+							interest_rate,
+							reference_date.clone(),
+						)?
+						.iter()
+						// 1 payment every 6 months
+						.step_by(6)
```

Comment:
Why `SemiAnually` is not implemented, calling to some `get_semianually_cashflows()` and instead, it is built composing six `get_monthly_cashflows()`?. 

I mean, is it truly composable? Seems like the `reference_rate` will be computed monthly instead of semiannually, isn't it?

Reply:
No you're right, the reference dates would be incorrect. Would be nice to re-use some logic but I think it's indeed not really feasible.